File: app/src/main/java/com/lineclient/home/homelineclient/ws/WsService.java
package com.lineclient.home.homelineclient.ws;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import com.lineclient.home.homelineclient.net.NetDataConstants;
import com.lineclient.home.homelineclient.tools.Debug;

import org.java_websocket.client.WebSocketClient;
import org.java_websocket.drafts.Draft;
import org.java_websocket.drafts.Draft_17;
import org.java_websocket.handshake.ServerHandshake;

import java.lang.ref.WeakReference;
import java.net.URI;
import java.net.URISyntaxException;

/**
 * Created by yangfengyuan on 16/7/28.
 */
public class WsService extends Service {

    private boolean isToTryAuto = true;
    private WsGetDataInterface wsGetDataInterface;
    private URI serviceUrl;
    private WebSocketWorker webSocketWorker;
    private boolean isConnect = false;
    private MsgHandler msgHandler = new MsgHandler(this);

    public interface WsGetDataInterface {

        void onClose();

        void onError(String error);

        void onMessage(String msg);

        void onOpen();

    }

    public class LocalBinder extends Binder {
        public WsService getService() {
            return WsService.this;
        }
    }

    public boolean isConnect() {
        return isConnect;
    }

    public boolean isToTryAuto() {
        return isToTryAuto;
    }

    public void setToTryAuto(boolean toTryAuto) {
        isToTryAuto = toTryAuto;
    }

    public void setWsGetDataInterface(WsGetDataInterface wsGetDataInterface) {
        this.wsGetDataInterface = wsGetDataInterface;
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {

        try {

            serviceUrl = new URI(NetDataConstants.WS_URL);

            startWs();

        } catch (URISyntaxException e) {

            e.printStackTrace();

        }

        return new LocalBinder();

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
    }

    @Override
    public boolean onUnbind(Intent intent) {
        return super.onUnbind(intent);
    }

    private void startWs() {

        new Thread(new Runnable() {
            @Override
            public void run() {

                connectWSService();

            }
        }).start();

    }

    public boolean retryConnect() {

        try {
            if (!webSocketWorker.connectBlocking()) {
                webSocketWorker.connect();
            } else {
                return false;
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return false;

    }

    public void connectWSService() {
        webSocketWorker = new WebSocketWorker(serviceUrl, new Draft_17());
        webSocketWorker.connect();
    }

    public void closeConnext() {
        webSocketWorker.close();
    }

    public void sendData(String msg) {
        webSocketWorker.send(msg);
    }

    private static class MsgHandler extends Handler {

        WeakReference<WsService> weakReference;

        MsgHandler(WsService object) {
            weakReference = new WeakReference<>(object);
        }

        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);

            WsService object = weakReference.get();
            if (object != null) {
                switch (msg.what) {
                    case 0:
                        object.wsGetDataInterface.onClose();
                        break;
                    case 1:
                        object.wsGetDataInterface.onOpen();
                        break;
                    case 2:
                        object.wsGetDataInterface.onError((String) msg.obj);
                        break;
                    case 3:
                        String data = (String) msg.obj;
                        object.wsGetDataInterface.onMessage(data);
                        break;
                }
            }

        }

    }

    private class WebSocketWorker extends WebSocketClient {

        public WebSocketWorker(URI serverUri, Draft draft) {
            super(serverUri, draft);
        }

        @Override
        public void onClose(int arg0, String arg1, boolean arg2) {
            msgHandler.sendEmptyMessage(0);
            isConnect = false;
        }

        @Override
        public void onError(Exception error) {

            isConnect = false;

            String errorMsg = error.getMessage();

            if (TextUtils.isEmpty(errorMsg)) {
                errorMsg = "未知连接错误";
            }

            Message msg = new Message();
            msg.what = 2;
            msg.obj = errorMsg;
            msgHandler.sendMessage(msg);

            if (isToTryAuto) {

                try {
                    Thread.sleep(5000);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }

                if (!isConnect) {

                    connectWSService();
                }

            }

        }

        @Override
        public void onMessage(String data) {
            Message msg = new Message();
            msg.what = 3;
            msg.obj = data;
            msgHandler.sendMessage(msg);
        }

        @Override
        public void onOpen(ServerHandshake arg0) {
            msgHandler.sendEmptyMessage(1);
            isConnect = true;
        }

    }
}

File: app/src/main/java/com/lineclient/home/homelineclient/ws/WsService.kt
package com.lineclient.home.homelineclient.ws

import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.Handler
import android.os.IBinder
import android.os.Message
import android.text.TextUtils
import com.lineclient.home.homelineclient.net.NetDataConstants
import com.lineclient.home.homelineclient.ws.WsService
import org.java_websocket.client.WebSocketClient
import org.java_websocket.drafts.Draft
import org.java_websocket.drafts.Draft_17
import org.java_websocket.handshake.ServerHandshake
import java.lang.ref.WeakReference
import java.net.URI
import java.net.URISyntaxException

/**
 * Created by yangfengyuan on 16/7/28.
 */
class WsService : Service() {
    var isToTryAuto = true
    private var wsGetDataInterface: WsGetDataInterface? = null
    private var serviceUrl: URI? = null
    private var webSocketWorker: WebSocketWorker? = null
    var isConnect = false
        private set
    private val msgHandler = MsgHandler(this)

    interface WsGetDataInterface {
        fun onClose()
        fun onError(error: String?)
        fun onMessage(msg: String?)
        fun onOpen()
    }

    inner class LocalBinder : Binder() {
        val service: WsService
            get() = this@WsService
    }

    fun setWsGetDataInterface(wsGetDataInterface: WsGetDataInterface?) {
        this.wsGetDataInterface = wsGetDataInterface
    }

    override fun onBind(intent: Intent): IBinder? {
        try {
            serviceUrl = URI(NetDataConstants.WS_URL)
            startWs()
        } catch (e: URISyntaxException) {
            e.printStackTrace()
        }
        return LocalBinder()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    override fun onUnbind(intent: Intent): Boolean {
        return super.onUnbind(intent)
    }

    private fun startWs() {
        Thread(Runnable { connectWSService() })
                .start()
    }

    fun retryConnect(): Boolean {
        try {
            if (webSocketWorker?.connectBlocking() == false) {
                webSocketWorker?.connect()
            } else {
                return false
            }
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        return false
    }

    fun connectWSService() {
        webSocketWorker =
            WebSocketWorker(serviceUrl, Draft_17())
        webSocketWorker?.connect()
    }

    fun closeConnext() {
        webSocketWorker?.close()
    }

    fun sendData(msg: String?) {
        if (isConnect) {
            webSocketWorker?.send(msg)
        }
    }

    private class MsgHandler internal constructor(objects: WsService) : Handler() {
        var weakReference: WeakReference<WsService> = WeakReference(objects)
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            val objects = weakReference.get()
            when (msg.what) {
                0 -> objects?.wsGetDataInterface?.onClose()
                1 -> objects?.wsGetDataInterface?.onOpen()
                2 -> objects?.wsGetDataInterface?.onError(msg.obj as String)
                3 -> {
                    val data = msg.obj as String
                    objects?.wsGetDataInterface?.onMessage(data)
                }
            }
        }

    }

    private inner class WebSocketWorker(
        serverUri: URI?,
        draft: Draft?
    ) : WebSocketClient(serverUri, draft) {
        override fun onClose(
            arg0: Int,
            arg1: String,
            arg2: Boolean
        ) {
            msgHandler.sendEmptyMessage(0)
            isConnect = false
        }

        override fun onError(error: Exception) {
            isConnect = false
            var errorMsg = error.message
            if (TextUtils.isEmpty(errorMsg)) {
                errorMsg = "未知连接错误"
            }
            val msg = Message()
            msg.what = 2
            msg.obj = errorMsg
            msgHandler.sendMessage(msg)
            if (isToTryAuto) {
                try {
                    Thread.sleep(5000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
                if (!isConnect) {
                    connectWSService()
                }
            }
        }

        override fun onMessage(data: String) {
            val msg = Message()
            msg.what = 3
            msg.obj = data
            msgHandler.sendMessage(msg)
        }

        override fun onOpen(arg0: ServerHandshake) {
            msgHandler.sendEmptyMessage(1)
            isConnect = true
        }
    }
}